adding upgrade guide

nlMeminger opened the pull request:
This PR adds a guide to upgrading the application.  This is mostly based on my discussion with @jwflory, with some extra information given on database concerns during upgrades. Closes #192

Commit:
adding upgrade guide

leong96 commented:
@nlMeminger Guide seems okay, but I'll let Justin look it over. However, it's time to be a grammar nazi (or something?).

- 1. if a new version were released -> 1. If a new version was released.
- 2. you have new features you have developed that you would like to move into production. -> 2. You have new features you have developed that you would like to move into production.
- A period at the end of the note in line 7.
- On line 10, I think we could go into some short note on what cases the database data will get erased. Last I checked, changing the database schema would not erase data, unless the user really screwed up (or they removed a column from models.py, a possibility I just thought of now). It's a good idea to warn the user to make a backup of the database, but claiming "this will most likely" without any further explanation seems like we're leaving the admin hanging. What should the user do then to minimize the chances of data loss?
- In line 13, Grasa Locator App -> GRASA Events Locator App. We should get the name of the client company (and the name of our own application) right at least. Just a matter of consistency.
- In line 15, is it zipfile, zip file, or .zip file?
- In line 20, why is Container capitalized?
- In line 24, the start -> then start.

That's all I got.

nlMeminger commented:
Reading your comments through email without the doc in front of me, but
1. It is “if a ...were..” in this case because it’s being used in a
conditional sentence about a possible but not necessarily true future. I’ll
fight you on this one.
2. I don’t see the difference in your two sentences
3. Will add a period
4. You and Eli know more about the specifics of this. If you could provide
a better explanation I would appreciate it.
5. I will look at how it’s used in other docs, and change accordingly
6. Does it matter if it’s zipfile zip file or .zip file? Though, as I’m
typing this on my phone, autocorrect did not like zipfile so...
7 & 8 good catches, will fix those

On Tue, Nov 26, 2019 at 00:05 leong96 <notifications@github.com> wrote:

> @nlMeminger <https://github.com/nlMeminger> Guide seems okay, but I'll
> let Justin look it over. However, it's time to be a grammar nazi (or
> something?).
>
>    -
>       1. if a new version were released -> 1. If a new version was
>       released.
>    -
>       1. you have new features you have developed that you would like to
>       move into production. -> 2. You have new features you have developed that
>       you would like to move into production.
>    - A period at the end of the note in line 7.
>    - On line 10, I think we could go into some short note on what cases
>    the database data will get erased. Last I checked, changing the database
>    schema would not erase data, unless the user really screwed up (or they
>    removed a column from models.py, a possibility I just thought of now). It's
>    a good idea to warn the user to make a backup of the database, but claiming
>    "this will most likely" without any further explanation seems like we're
>    leaving the admin hanging. What should the user do then to minimize the
>    chances of data loss?
>    - In line 13, Grasa Locator App -> GRASA Events Locator App. We should
>    get the name of the client company (and the name of our own application)
>    right at least. Just a matter of consistency.
>    - In line 15, is it zipfile, zip file, or .zip file?
>    - In line 20, why is Container capitalized?
>    - In line 24, the start -> then start.
>
> That's all I got.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jwflory/django-rit-grasa/pull/249?email_source=notifications&email_token=AELFBENXYX4AJVCYO7SMKULQVSVAPA5CNFSM4JRSF5M2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEFEXG2Q#issuecomment-558461802>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AELFBELSUTF7RWMWEM3UI6LQVSVAPANCNFSM4JRSF5MQ>
> .
>

leong96 commented:
Thanks for the reply, yeah, let's see here.

1. I reread the sentence with your comments. In hindsight, were released actually does sound right, not sure why I thought that was grammatically incorrect. You can have that one (plus I'm in bed, why put up a fight?). The period (or an "or ;" at the end to lead to the next item, something to have some punctuation at the end of that) at the end still stands, basic enough.
2. Should have specified there; the difference is just a capital Y after the "2.". Now whether this affects how the page looks stylistically, I don't know, don't really have the actual docs webpage up, just the .md file, I'm thinking capitalizing the first letter in those list entries looks more right. Or at the least, if you could tell me what you were thinking of by making the list items lowercase.
3. Cool~.

4. Hopefully I remember this correctly, due to changing the schema once or twice earlier in development. Messing with the models can happen in four ways, that will result in either data loss, no change, or the addition of data:
    - Deleting a model.
         If the user is doing this, there's guaranteed data loss for the obvious reason; you're deleting a table that had app data; that's the reason the table was there in the first place.
    - Adding a model.
        There's no data loss from this; adding a table shouldn't affect the app's behavior, at least assuming no changes to the code.
    - Removing columns from an existing table.
         Obvious data loss here, a whole column is being removed.
   - Adding columns to an existing table.
         Adding some types of columns to a table will force you to also specify a default value (ie contact_name = models.CharField(max_length=255, default="Name not Specified")). Some will not (ie models.TextField). No data is lost, though if a column that requires a default is added to a table, existing entries will now have that default value in the column post-migration.

Based on the above, I'd imagine the most typical type of schema change (adding columns for new purposes, rather than removing the existing columns) would most likely not result in data loss. And if a developer in the future were to delete an existing column, they'd better know that there would be data loss. Thinking about it, as for how this fits in the docs, think the problem I had is that there's a good chance there's won't be data loss, so maybe change "will most likely" -> "can potentially", basically something that doesn't automatically imply a migration will wipe your data. Because in most cases, it probably won't. Or if you want to keep the will most likely wording, you can then add that it will if one of the above scenarios I described above happens. @R-Eli-Levasseur may be able to word it better as well.

5. Please do, yes. At least change Grasa to GRASA; that's the actual organization name. :D

6. I leave you to determine the zip file stuff~ The wording just got to me because I've seen the zipfile wording used informally on occasion, but these are docs to be read by some employee of Monroe County, might as well be correct in a formal context with wording.

7/8. Cool~

Fight-o! And good night.

justwheel commented on docs/admin/upgrade.md:
```diff
@@ -0,0 +1,27 @@
+
+## Database Concerns
+It is _highly_ recommended to make a backup of your database before making any upgrades.  If the new version you are upgrading to has database schema changes, then you will need to upgrade your migrations file and remigrate your database.  This will most likely wipe your data.  It is recommended to start a new database alongside your current database and update your config file with the new datab ...
+## How To Actually Upgrade
```
You need a newline before the beginning of a header for it to render correctly in the docs:

```suggestion

 ## How To Actually Upgrade 
```

justwheel commented on docs/admin/upgrade.md:
```diff
@@ -0,0 +1,27 @@
+It is _highly_ recommended to make a backup of your database before making any upgrades.  If the new version you are upgrading to has database schema changes, then you will need to upgrade your migrations file and remigrate your database.  This will most likely wipe your data.  It is recommended to start a new database alongside your current database and update your config file with the new datab ...
+## How To Actually Upgrade
+
+The upgrade process is rather simple.  The steps for upgrading to a new version of the Grasa Locator App are as follows:
```
I agree with @leong96 to be consistent on the name and spelling of the app:

```suggestion
The upgrade process is rather simple.
The steps for upgrading to a new version of the GRASA Event Locator are as follows:
```

justwheel commented on docs/admin/upgrade.md:
````diff
@@ -0,0 +1,27 @@
+2. Stop the current running application by running
+    ```sh
+    ./up.py --stop
+    ```
````
In case volumes are ever used, this should be the first step before copying new code into the application folder while it's running. If there is a volume, it will copy the new code into a live environment even if there are important back-end / DB changes not yet updated.

Please swap order of Steps 1 and 2.

justwheel reviewed (changes requested):
Thanks for working on this @nlMeminger. I left some feedback in-line, and in most cases I left interactive suggestions for one-click fixes.

However, before merging, I would like you to please write each sentence on a new line ([why? explained here](https://asciidoctor.org/docs/asciidoc-recommended-practices/#one-sentence-per-line)). This will help with git diffs and since I plan to make style changes to this doc after it is merged anyways, it will make my changes easier to review if each sentence is written on a new line.

Once this is done and other feedback is addressed, this should be ready to go.

nlMeminger committed:
Apply suggestions from code review

applying some of @jwflory's suggestions

Co-Authored-By: Justin W. Flory <4721034+jwflory@users.noreply.github.com>

nlMeminger committed:
Update upgrade.md

Making changes to the upgrade doc in accordence with @jwflory's and @leong96's feedback.

nlMeminger commented:
@jwflory, thanks for the feedback.  I've made some changes that take into account both your and @leong96's feedback.  Another review would be appreciated.

nlMeminger committed:
adding newlines

adding more newlines.  I think I've got everything now...

justwheel reviewed (approved):
Thanks @nlMeminger, these changes look good. I'm going to go ahead and merge this. :ocean:
